website/blog/deployers.md

```diff
@@ -0,0 +1,525 @@
+responds with a [ReplicaToRegister][ReplicaToRegister] protobuf with information
+about itself (e.g., its address). Here's an example of what the `WeaveletMsg`
+looks like when printed by our deployer:
+
+```json
+{
+    "id": "1",
+    "replicaToRegister": {
+        "address":      "tcp://100.116.161.153:35077",
+    }
+}
+```
+
+This trivial deployer demonstrates how a deployer and weavelet exchange
+information via protobufs over a pair of pipes. Understanding these low-level
+details is important if you want to implement a deployer completely
+scratch, say in another programming language. For even more details, refer to
+[runtime.proto][].
+
+## Envelopes
+
+Service Weaver provides an `Envelope` API that abstracts away all the tedious
+code we wrote by hand in the previous section. An `Envelope` runs a Service
+Weaver binary as a subprocess; establishes pipes to the weavelet running in the
+subprocess; and serializes and deserializes protos over these pipes.
+
+We construct an `Envelope` with an instance of the [EnvelopeHandler][]
+interface. This interface defines how a deployer handles requests from the
+weavelet.
+
+```go
+type EnvelopeHandler interface {
+    // Responsibility 1: Telemetry.
+    RecvLogEntry(*protos.LogEntry)
+    RecvTraceSpans([]trace.ReadOnlySpan) error
+
+    // Responsibility 2: Listeners.
+    GetAddress(*protos.GetAddressRequest) (*protos.GetAddressReply, error)
+    ExportListener(*protos.ExportListenerRequest) (*protos.ExportListenerReply, error)
+
+    // Responsibility 3: Components.
+    StartComponent(*protos.StartComponentRequest) error
+}
+```
+
+The interface includes methods to implement a deployer's three responsibilities:
+telemetry, listeners, and components. We'll explain these methods in depth in
+the next section when we implement a deployer.
+
+An `Envelope` also includes methods that we can call to send requests to the
+weavelet. Again, we can divide these methods based on a deployer's three
+responsibilities. We'll also explain these methods in depth in the next section
+when we implement a deployer.
+
+```go
+// Responsibility 1: Telemetry.
+Envelope.GetHealth(context.Context) protos.HealthStatus
+Envelope.GetProfile(context.Context, req *protos.GetProfileRequest) (*protos.Profile, error)
+Envelope.GetMetrics(context.Context) ([]*metrics.MetricSnapshot, error)
+
+// Responsibility 3: Components.
+Envelope.UpdateRoutingInfo(context.Context, *protos.RoutingInfo) error
+Envelope.UpdateComponents(context.Context, []string) error
+```
+
+## A Simple Multiprocess Deployer
+
+In this section, we implement a fully working multiprocess deployer. To make
+things simple, our deployer won't co-locate or replicate any components. Every
+component will run by itself in a single weavelet. See [here][multi_deployer]
+for the full source code.
+
+We begin by declaring types for the deployer, a weavelet, and an
+`EnvelopeHandler` implementation.
+
+```go
+package main
+
+import ...
+
+type deployer struct {
```

Comment:
Add a comment inside the struct:

This deployer gives each component its own weavelet.